### PR TITLE
Accept Feature arg

### PR DESCRIPTION
**Pull request by jonpsy**
Fix so that get_log_density() works even after passing Feature.  @vigsterkr , comments?

### CODE REVIEW

File: src/shogun/distributions/KernelDensity.cpp [hunk 1]
```diff
@@ -91,12 +91,14 @@ bool KernelDensity::train(std::shared_ptr<Features> data)
 	return true;
 }
 
-SGVector<float64_t> KernelDensity::get_log_density(const std::shared_ptr<DenseFeatures<float64_t>>& test, int32_t leaf_size)
+SGVector<float64_t> KernelDensity::get_log_density(const std::shared_ptr<Features>& test, int32_t leaf_size)
 {
 	require(test,"data not supplied");
-
+	require(test->get_feature_type == 10 && test->get_feature_class == 11,"Expecting DenseFeatures<float64_t>");
+	std::shared_ptr<DenseFeatures<float64_t>> dense_feat = std::static_pointer_cast<DenseFeatures<float64_t>>(test);
```
Comment (gf712) [inline src/shogun/distributions/KernelDensity.cpp]: imo you should do a dynamic cast and check for nullptr, you shouldn't rely on `get_feature_type` that might get deprecated soon.

Comment (jonpsy) [reply src/shogun/distributions/KernelDensity.cpp]: Aight I'll update that

Comment (jonpsy) [reply src/shogun/distributions/KernelDensity.cpp]: I've used ```dynamic_pointer_cast``` as it's ```std::shared_ptr``` thoughts?

File: src/shogun/distributions/KernelDensity.cpp [hunk 2]
```diff
@@ -91,12 +91,14 @@ bool KernelDensity::train(std::shared_ptr<Features> data)
 	return true;
 }
 
-SGVector<float64_t> KernelDensity::get_log_density(const std::shared_ptr<DenseFeatures<float64_t>>& test, int32_t leaf_size)
+SGVector<float64_t> KernelDensity::get_log_density(const std::shared_ptr<Features>& test, int32_t leaf_size)
 {
 	require(test,"data not supplied");
-
+	std::shared_ptr<DenseFeatures<float64_t>> dense_feat =std::dynamic_pointer_cast<DenseFeatures<float64_t>>(test);
```
Comment (gf712) [inline src/shogun/distributions/KernelDensity.cpp]: `auto dense_feat = ...` you already show what the type is in the cast

Comment (jonpsy) [reply src/shogun/distributions/KernelDensity.cpp]: True

File: src/shogun/distributions/KernelDensity.cpp (hunk 1 above)
Comment (gf712) [reply src/shogun/distributions/KernelDensity.cpp]: yes that's what you should use

File: src/shogun/distributions/KernelDensity.cpp (hunk 2 above)
Comment (jonpsy) [reply src/shogun/distributions/KernelDensity.cpp]: Done

File: src/shogun/distributions/KernelDensity.cpp [hunk 3]
```diff
@@ -91,12 +91,14 @@ bool KernelDensity::train(std::shared_ptr<Features> data)
 	return true;
 }
 
-SGVector<float64_t> KernelDensity::get_log_density(const std::shared_ptr<DenseFeatures<float64_t>>& test, int32_t leaf_size)
+SGVector<float64_t> KernelDensity::get_log_density(const std::shared_ptr<Features>& test, int32_t leaf_size)
```
Comment (karlnapf) [inline src/shogun/distributions/KernelDensity.cpp:94]: I don't really see the point of moving this check to runtime?

Comment (vigsterkr) [reply src/shogun/distributions/KernelDensity.cpp:94]: @karlnapf unified/simplified API ;)

Comment (vigsterkr) [reply src/shogun/distributions/KernelDensity.cpp:94]: features() returns Features... you wont be able to call this from SWIG API, as you dont have RealFeatures anymore...

Comment (karlnapf) [reply src/shogun/distributions/KernelDensity.cpp:94]: sure thing. I didnt realise this is exposed to swig.

Comment (vigsterkr) [reply src/shogun/distributions/KernelDensity.cpp:94]: well KDE notebook uses it :S

Comment (vigsterkr) [reply src/shogun/distributions/KernelDensity.cpp:94]: i guess there could be a way to refactor this even more so the function is not exposed at all, but on the end somehow features should get to this object, i.e. it'll have to be general Features as well...

Comment (karlnapf) [reply src/shogun/distributions/KernelDensity.cpp:94]: yeah, the public interface needs the base class for sure. I think this is good in the meantime and at some point we could/should refactor the distribution api....
Also with this, we can add a meta example for distributions!

**Comment by gf712** [comment]
can you trigger the CI again to see if everything is working please?

**Pull request merged**